[protocol] commit-reveal scheme for non-reverted randomness (#123)
onflow/flips: protocol/20230728-commit-reveal.md

### Performance Implications

- There is no timing and gas impact other then the PRG is being executed on-chain.
- The core-contract will have to store the SoR history for a limited window. For `N` equal to 1200000, the storage size is at least 37Mb (not counting the encoding overhead).

### Dependencies

None

### Engineering Impact

The points 1, 2 and 3 in [the design proposal](#design-proposal) need to be built as part of this proposal.

### Tutorials and Examples

Here is an example of a coin toss contract with one function to commit a bid, and another function to resolve the bid:

```
import SoRHistory from 0xFLOWSORHISTORY

cointossCommit(bet: @FlowToken.Vault) : @Receipt {

> 		let receipt <- create Receipt(

Reviewer: Almost feels like this Receipt object should have a "prg" function that is call-able, is that the intention?

Author: There are many ways are possible to organize the code. Here, I wanted to explicitly call `SoRHistory.getSoR` to show how to query the history SoR core-contract, and use it to seed a PRG.

					betAmount: bet.balance,
					commitBlock: getCurrentBlock().height
		)
		self.reserve.deposit(from: <-bet)
    return <- receipt
}

cointossReveal(receipt: @Receipt) : @FlowToken.Vault {
		let currentBlock = getCurrentBlock().height
		if receipt.commitBlock >= currentBlock {
			panic("cannot reveal yet")
		}

>         // `expiryWindowLength` is optionally defined by the application to be less than `N`.

Reviewer: seems to be some spacing issues with the example code?

        //  if not `N` would be implicitly used as an expiry window.
        const let expiryWindowLength = 1000000
        if receipt.commitBlock + expiryWindowLength > currentBlock {
            return <-FlowToken.createEmptyVault()
        }
		
		let coin = randomCoin(atBlock: receipt.commitBlock, salt: receipt.id)
		destroy receipt

		if coin == 1 {
			return <-FlowToken.createEmptyVault()
		}
		
		return <-self.reserve.withdraw(amount: receipt.betAmount * 2)
}

fun randomCoin(atBlock: UInt64, salt: UInt64) : UInt8 {

>     let sor = SoRHistory.getSoR(atBlock)

Reviewer: this SOR is exposed to the user/dapp developer? Any concerns with that?

Author: I don't think so. That is the SoR of a past block. The current block's SoR isn't available in the contract `SoRHistory`.

Reviewer 2: Maybe document in the FLIP that only past SoR are available

### Performance Implications

- There is no timing and gas impact other then the PRG is being executed on-chain.
- The core-contract will have to store the SoR history for a limited window. For `N` equal to 1200000, the storage size is at least 37Mb (not counting the encoding overhead).

### Dependencies

None

### Engineering Impact

The points 1, 2 and 3 in [the design proposal](#design-proposal) need to be built as part of this proposal.

### Tutorials and Examples

Here is an example of a coin toss contract with one function to commit a bid, and another function to resolve the bid:

```

> import SoRHistory from 0xFLOWSORHISTORY
> import PRG from 0xPRGIMPLEMENTATION

Reviewer 2: Tooling convention is to use string literal imports, which get rewritten to the actual address depending on network:

```suggestion
import SoRHistory from "SoRHistory"
import PRG from "PRG"
```

### Performance Implications

- There is no timing and gas impact other then the PRG is being executed on-chain.
- The core-contract will have to store the SoR history for a limited window. For `N` equal to 1200000, the storage size is at least 37Mb (not counting the encoding overhead).

### Dependencies

None

### Engineering Impact

The points 1, 2 and 3 in [the design proposal](#design-proposal) need to be built as part of this proposal.

### Tutorials and Examples

Here is an example of a coin toss contract with one function to commit a bid, and another function to resolve the bid:

```
import SoRHistory from 0xFLOWSORHISTORY

> cointossCommit(bet: @FlowToken.Vault) : @Receipt {

Reviewer 2: ```suggestion
fun commitCointoss(bet: @FlowToken.Vault): @Receipt {
```

	let receipt <- create Receipt(
				betAmount: bet.balance,
				// commit to use randomness at the current block (still unknown)
				commitBlock: getCurrentBlock().height
	)
	// commit the bet

> 	self.reserve.deposit(from: <-bet)

Reviewer 2: What is `self.reserve`?

Author: @janezpodhostnik can you help me define `reserve` in the example please?

Reviewer 3: Its basicaly the main pool of funds of the coin-toss contract. As it is written in the example it would be a `@FlowToken.Vault` field on the contract itself. It would be seeded with some initial money so that the contract could function.

Reviewer 2: Thank you for the great explanation? Could you please add this information somehow to the FLIP itself? For example, maybe add comments to the code and/or the explanation as text next to the code block.

Author: I added a comment to the example (https://github.com/onflow/flips/pull/123/commits/c648ffa0a5bb09341d2305143a550eb2687bc84d) but Janez feel free to add a proper field if you prefer.

Reviewer 3: I don't think Its necessary to add the field to the cadence example. I think its better to keep the focus on the usage and not on how this theoretical contract would function, but let me know if you disagree.

    return <- receipt
}

> cointossReveal(receipt: @Receipt) : @FlowToken.Vault {

Reviewer 2: ```suggestion
fun revealCointoss(receipt: @Receipt): @FlowToken.Vault {
```

	let currentBlock = getCurrentBlock().height
	if receipt.commitBlock >= currentBlock {
		panic("cannot reveal yet")
	}
	// `expiryWindowLength` is optionally defined by the application to be less than `N`.
	//  if not `N` would be implicitly used as an expiry window.

> 	const let expiryWindowLength = 1000000

Reviewer 2: ```suggestion
	let expiryWindowLength = 1000000
```

	if receipt.commitBlock + expiryWindowLength > currentBlock {

> 		return <-FlowToken.createEmptyVault()

Reviewer 2: `receipt` is lost / not properly invalidated. Maybe read `receipt.commitBlock` into a temporary, perform the `destroy receipt`, then have this if block here.

	}
	
	let coin = randomCoin(atBlock: receipt.commitBlock, salt: receipt.id)
	destroy receipt

	if coin == 1 {
		return <-FlowToken.createEmptyVault()
	}
	
	return <-self.reserve.withdraw(amount: receipt.betAmount * 2)
}

> fun randomCoin(atBlock: UInt64, salt: UInt64) : UInt8 {

Reviewer 2: ```suggestion
fun randomCoin(atBlock: UInt64, salt: UInt64): UInt8 {
```

	// query the SoR history core-contract
	let sor = SoRHistory.getSoR(atBlock)
	// instantiate a PRG

> 	var prg = createPRG(sor, salt)
> 	let rand = prg.Uint64()

Reviewer 2: Maybe define `createPRG` and the `PRG` type, incl. the `Uint64` function, and leave the implementation out. 

Also, the variable is never re-assigned, so can be `let`
```suggestion
	let prg = createPRG(sor, salt)
	let rand = prg.Uint64()
```

Author: I was assuming the `PRG` object and its method are defined by another contract (which is supposed to be implemented as part of this proposal). 
I imported another contract in the example and added comments to clarify, but happy to make further adjustments.

> 	return rand & 1

Reviewer 2: Need to explicitly convert `UInt64` to return type `UInt8`:
```suggestion
	return UInt8(rand & 1)
```

## Design Proposal

The proposed design is a commit-reveal pattern. 

The solution requires infrastructure changes to provide new data to the transaction execution environment:

 1. a new FVM function that exposes the current block's SoR (more precisely a derived value from the protocol `SoR_A`) to Cadence runtime. Note that `unsafeRandom` only exposes randoms derived from SoR through a pseudo-random generator (PRG) but not the SoR itself. 
 2. a new system core-contract that stores a limited history of SoRs for the past `N` blocks. The new FVM function in (1) is only available to the history contract, and is not available to other non-system transactions. Note that system-transactions in Flow are executed at the end of each block, after all non-system transactions of the block are executed. The proposal suggests to change the system transaction so that it adds the current block's `SoR_A` to the SoR history contract (it also removes the oldest block's `SoR_A`). The contract indexes the SoR history by block height. 

>  3. an on-chain implementation of a PRG is required. A PRG is initialized with an SoR and can generate a sequence of random numbers for an application. It's up to the application to use a suitable PRG instance. At least one recommended PRG implementation should be provided as part of the FLIP implementation (precise instance to be determined).

Reviewer 4: Curious of the status and anticipated effort required for the recommended PRG implementation.

Author: The PRG used in the protocol is a ChaCha20 based PRG. This is a crypto-secure PRG that is not required for the on-chain case. It also might be too heavy to implement on-chain. 

A lighter version and easy to implement is XorShift128+. IMO, implementing this FLIP can include one PRG instance (for instance XorShift128+), but developers can implement other instances and are free to choose the PRG they want in their contracts.

Note that this is an inherent behavior of atomic contract platforms where the programming language is able to error and all temporary results of the transaction are not applied to to the state. While this is not a limitation of the programming language or the safe distributed randomness, a contract platform should provide developers with tools to implement a non-revertible usage of randomness. 

## User Benefit

Adding a safe pattern to reveal randomness without the transaction results being reverted unlocks application relying on randomness (for instance the coin toss contract described above). Not providing such pattern pushes developers to rely on risky solutions with possible security flaws.

## Design Proposal

The proposed design is a commit-reveal pattern. 

The solution requires infrastructure changes to provide new data to the transaction execution environment:

>  1. a new FVM function that exposes the current block's SoR (more precisely a derived value from the protocol `SoR_A`) to Cadence runtime. Note that `unsafeRandom` only exposes randoms derived from SoR through a pseudo-random generator (PRG) but not the SoR itself. 

Author: I think that's what is done in this [PR](https://github.com/onflow/flow-go/pull/4582): FVM exposes the source of randomness needed by the contract, and it is called in the service transaction at the end of the block.
Let me know if there is a conflict with what the FLIP describes.

>  1. a new FVM function that exposes the current block's SoR (more precisely a derived value from the protocol `SoR_A`) to Cadence runtime. Note that `unsafeRandom` only exposes randoms derived from SoR through a pseudo-random generator (PRG) but not the SoR itself. 

Reviewer 4: Ah right, my mistake. Found the method in the [system chunk transaction](https://github.com/onflow/flow-go/blob/1231763cb306d6d81cdf7c922c5761097d1dcfb9/fvm/blueprints/scripts/systemChunkTransactionTemplate.cdc#L19)

```
import RandomBeaconHistory from "RandomBeaconHistory"
import PRG from "PRG"
// PRG implementation is not provided by the FLIP, we assume this contract
// imports a suitable PRG implementation

fun commitCointoss(bet: @FlowToken.Vault): @Receipt {
	let receipt <- create Receipt(
				betAmount: bet.balance,
				// commit to use randomness at the current block (still unknown)
				commitBlock: getCurrentBlock().height
	)
	// commit the bet
	// `self.reserve` is a `@FlowToken.Vault` field defined on the app contract
	//  and represents a pool of funds
	self.reserve.deposit(from: <-bet)
    return <- receipt
}

fun revealCointoss(receipt: @Receipt): @FlowToken.Vault {
	let currentBlock = getCurrentBlock().height
	if receipt.commitBlock >= currentBlock {
		panic("cannot reveal yet")
	}
	
	let winnings = receipt.betAmount * 2
	let coin = randomCoin(atBlock: receipt.commitBlock, salt: receipt.id)
	destroy receipt

	if coin == 1 {
		return <-FlowToken.createEmptyVault()
	}
	
	// `self.reserve` is a `@FlowToken.Vault` field defined on the app contract
	//  and represents a pool of funds
	return <-self.reserve.withdraw(amount: winnings)
}

> fun randomCoin(atBlock: UInt64, salt: UInt64): UInt8 {
> 	// query the Random Beacon history core-contract.
> 	// if `atBlock` is the current block, `SourceOfRandomnessAtBlock` errors.
> 	let sor = RandomBeaconHistory.SourceOfRandomnessAtBlock(atBlock)
> 
> 	// instantiate a PRG object using external `createPRG` that takes a `seed` 
> 	// and `salt` and returns a pseudo-random-generator object.
> 	let prg = PRG.createPRG(sor, salt)
> 
> 	// derive a 64-bit random using the object `prg`
> 	let rand = prg.Uint64()
> 	return UInt8(rand & 1)
> }

Reviewer 2: ```suggestion
fun randomCoin(atBlockHeight blockHeight: UInt64, salt: UInt64): UInt8 {
	// query the Random Beacon history core-contract.
	// if `blockHeight` is the current block height, `sourceOfRandomness` errors.
	let sourceOfRandomness = RandomBeaconHistory.sourceOfRandomness(atBlockHeight: blockHeight)

	// instantiate a PRG object using external `createPRG` that takes a `seed` 
	// and `salt` and returns a pseudo-random-generator object.
	let prg = PRG.createPRG(sourceOfRandomness, salt)

	// derive a 64-bit random using the object `prg`
	let rand = prg.Uint64()
	return UInt8(rand & 1)
}
```
